codegen/templates.go

```diff
@@ -1,5 +1,258 @@
+				Schema:      schema,
+			}
+			result = append(result, toTemplatableParam(param, false))
+		}
+	}
+	return result
+}
+
+// templatableParam mainly just reuses the OASParameter,
+// but adds on a IsPathParam bool.
+type templatableParam struct {
+	*framework.OASParameter
+	IsPathParam bool
+}
+
+func toTemplatableParam(param framework.OASParameter, isPathParameter bool) *templatableParam {
+	ptrToParam := &param
+	if ptrToParam.Schema == nil {
+		// Always populate schema and display attributes so later it'll be easier
+		// to check if they're sensitive by iterating over them.
+		ptrToParam.Schema = &framework.OASSchema{}
+	}
+	if ptrToParam.Schema.DisplayAttrs == nil {
+		ptrToParam.Schema.DisplayAttrs = &framework.DisplayAttributes{}
+	}
+	return &templatableParam{
+		OASParameter: ptrToParam,
+		IsPathParam:  isPathParameter,
+	}
+}
+
+// templatableEndpoint is a convenience struct that plays nicely with Go's
+// template package. It is used to keep as much logic as possible in Go
+// rather than in Go's templating language, because most folks are more
+// familiar with Go.
+type templatableEndpoint struct {
+	Endpoint                string
+	DirName                 string
+	UpperCaseDifferentiator string
+	LowerCaseDifferentiator string
+	Parameters              []*templatableParam
+	SupportsRead            bool
+	SupportsWrite           bool
+	SupportsDelete          bool
+}
+
+func (e *templatableEndpoint) Validate() error {
+	if e == nil {
+		return fmt.Errorf("endpoint is nil")
+	}
+	if e.Endpoint == "" {
+		return fmt.Errorf("endpoint cannot be blank for %+v", e)
+	}
+	if e.DirName == "" {
+		return fmt.Errorf("dirname cannot be blank for %+v", e)
+	}
+	if e.UpperCaseDifferentiator == "" {
+		return fmt.Errorf("exported function prefix cannot be blank for %+v", e)
+	}
+	if e.LowerCaseDifferentiator == "" {
+		return fmt.Errorf("private function prefix cannot be blank for %+v", e)
+	}
+	for _, parameter := range e.Parameters {
+		isSupported := false
+		for _, supportedType := range supportedParamTypes {
+			if parameter.Schema.Type == supportedType {
+				if parameter.Schema.Type != "array" {
+					isSupported = true
+					break
+				}
+				// Right now, our templates assume that all array types are strings.
+				// If we allow other types of arrays, we will need to also go into
+				// each template and add additional logic supporting the new array
+				// type.
+				if parameter.Schema.Items.Type != "string" {
+					return fmt.Errorf("unsupported array type of %s for %s", parameter.Schema.Items.Type, parameter.Name)
+				}
+				isSupported = true
+				break
+			}
+		}
+		if !isSupported {
+			return fmt.Errorf("unsupported type of %s for %s", parameter.Schema.Type, parameter.Name)
+		}
+	}
+	return nil
+}
+
+// clean takes a field like "{role_name}" and returns
+// "rolename" for use in generated Go code.
+func clean(field string) string {
+	field = stripCurlyBraces(field)
+	return strings.Replace(field, "_", "", -1)
```

Comment:
That is added in PR 3 of this series (https://github.com/terraform-providers/terraform-provider-vault/pull/750/files#diff-4196ef9946b8ae0519ef95f84509ab20R255). Let's leave it there for now because the change still gets done, but doesn't require additional work.